pinot-core/src/main/java/com/linkedin/pinot/core/plan/TransformPlanNode.java

```diff
@@ -0,0 +1,171 @@
+
+    if (!transformExpressions.isEmpty()) {
+      _expressionTrees = buildTransformExpressionTrees(transformExpressions, true /* ignore columns */);
+      projectionColumns.addAll(getTransformColumns(_expressionTrees));
+    } else {
+      _expressionTrees = null;
+    }
+
+    _projectionPlanNode =
+        new ProjectionPlanNode(indexSegment, projectionColumns.toArray(new String[projectionColumns.size()]),
+            new DocIdSetPlanNode(indexSegment, brokerRequest));
+  }
+
+  /**
+   * Helper method to get all columns from expressions.
+   * @param expressionTrees List of expression trees for which to get the columns.
+   * @return List of columns from all the expression trees
+   */
+  private List<String> getTransformColumns(List<TransformExpressionTree> expressionTrees) {
+    List<String> columns = new ArrayList<>();
+    for (TransformExpressionTree expressionTree : expressionTrees) {
+      expressionTree.getColumns(columns);
+    }
+    return columns;
+  }
+
+  /**
+   * Helper method to build TransformExpressionTrees for a given set of transform expressions.
+   *
+   * @param transformExpressions List of expressions for which to build trees.
+   * @param ignoreColumns If true, expressions that are just column names are ignored.
+   * @return List of expression trees for the given transform expressions.
+   */
+  public static List<TransformExpressionTree> buildTransformExpressionTrees(List<String> transformExpressions,
+      boolean ignoreColumns) {
+    List<TransformExpressionTree> expressionTrees = new ArrayList<>(transformExpressions.size());
+
+    for (String transformExpression : transformExpressions) {
+      TransformExpressionTree expressionTree = _compiler.get().compileToExpressionTree(transformExpression);
+      if (!ignoreColumns || !expressionTree.isColumn()) {
+        expressionTrees.add(expressionTree);
+      }
+    }
+    return expressionTrees;
+  }
+
+  /**
+   * Helper method to extract projection columns and transform expressions from the given
+   * BrokerRequest.
+   *
+   * @param brokerRequest BrokerRequest to process
+   * @param projectionColumns Output projection columns from broker request
+   * @param transformExpressions Output transform expression from broker request
+   */
+  private void extractColumnsAndTransforms(BrokerRequest brokerRequest, List<String> projectionColumns,
+      List<String> transformExpressions) {
+
+    if (brokerRequest.isSetAggregationsInfo()) {
+      for (AggregationInfo aggregationInfo : brokerRequest.getAggregationsInfo()) {
+        if (!aggregationInfo.getAggregationType()
+            .equalsIgnoreCase(AggregationFunctionFactory.AggregationFunctionType.COUNT.getName())) {
+          String columns = aggregationInfo.getAggregationParams().get("column").trim();
+          projectionColumns.addAll(Arrays.asList(columns.split(",")));
+        }
+      }
+
+      // Collect all group by related columns.
+      if (brokerRequest.isSetGroupBy()) {
```

Comment:
The compile time for one entire query (not just expression) is about 0.2 ms on avg. Just because it shows up in the profiler does not mean it is expensive. This was already discussed in the team, I guess we can discuss this again.